src/hotspot/share/gc/shenandoah/shenandoahMark.inline.hpp

```diff
@@ -278,27 +278,24 @@ inline void ShenandoahMark::mark_through_ref(T *p, ShenandoahObjToScanQueue* q,
   if (!CompressedOops::is_null(o)) {
     oop obj = CompressedOops::decode_not_null(o);
 
+    ShenandoahHeap* heap = ShenandoahHeap::heap();
     shenandoah_assert_not_forwarded(p, obj);
-    shenandoah_assert_not_in_cset_except(p, obj, ShenandoahHeap::heap()->cancelled_gc());
+    shenandoah_assert_not_in_cset_except(p, obj, heap->cancelled_gc());
     if (in_generation<GENERATION>(obj)) {
       mark_ref<STRING_DEDUP>(q, mark_context, weak, obj);
       shenandoah_assert_marked(p, obj);
-      if (ShenandoahHeap::heap()->mode()->is_generational()) {
+      if (heap->mode()->is_generational()) {
         // TODO: As implemented herein, GLOBAL collections reconstruct the card table during GLOBAL concurrent
         // marking. Note that the card table is cleaned at init_mark time so it needs to be reconstructed to support
         // future young-gen collections.  It might be better to reconstruct card table in
         // ShenandoahHeapRegion::global_oop_iterate_and_fill_dead.  We could either mark all live memory as dirty, or could
         // use the GLOBAL update-refs scanning of pointers to determine precisely which cards to flag as dirty.
-        //
-        if ((GENERATION == YOUNG) && ShenandoahHeap::heap()->is_in(p) && ShenandoahHeap::heap()->is_in_old(p)) {
-          RememberedScanner* scanner = ShenandoahHeap::heap()->card_scan();
+        if (GENERATION == YOUNG && heap->is_in_old(p)) {
```

Comment:
I think I had added the heap->is_in(p) test because some program was sending in a non-heap value which was causing an assertion failure in is_in_old(p) during DEBUG builds (that have assertions enabled).  Maybe need to test this patch with FAST_DEBUG just to be sure we're ok with code as now written.

Reply:
I changed the implementation of `is_in_old` to first test for `is_in` in an earlier patch.